apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/HiPerfPublisherTest.cs

```diff
@@ -0,0 +1,225 @@
+        {
+            return new HiPerfPublisher.Settings(clients)
+            {
+                Scheduler = scheduler,
+                BatchingSettings = new BatchingSettings
+                {
+                    DelayThreshold = TimeSpan.FromMilliseconds(10000),
+                    ElementCountThreshold = batchElementCountThreshold,
+                    RequestByteThreshold = batchRequestByteThreshold
+                }
+            };
+
+        }
+
+        [Theory, CombinatorialData]
+        public void OneMessage(
+            [CombinatorialValues(false, true)] bool hardStop)
+        {
+            var topicName = new TopicName("FakeProject", "FakeTopic");
+            var scheduler = new TestScheduler();
+            TaskHelper taskHelper = scheduler.TaskHelper;
+            var client = new FakePublisher(scheduler, taskHelper);
+            var settings = MakeSettings(new[] { client }, scheduler);
+            int shutdownCount = 0;
+            var pub = new HiPerfPublisherImpl(topicName, settings, () =>
+            {
+                shutdownCount += 1;
+                return Task.FromResult(0);
+            }, taskHelper);
+            scheduler.Run(async () =>
+            {
+                var id = await taskHelper.ConfigureAwait(pub.PublishAsync("1"));
+                bool isCancelled = await taskHelper.ConfigureAwaitHideCancellation(pub.ShutdownAsync(new CancellationToken(hardStop)));
+                Assert.Equal(hardStop, isCancelled);
+                Assert.Equal("1", id);
+                Assert.Equal(1, shutdownCount);
+            });
+        }
+
+        [Theory, PairwiseData]
+        public void ManyMessages(
+            [CombinatorialValues(false, true)] bool hardStop,
+            [CombinatorialValues(1, 2, 5, 7)] int clientCount,
+            [CombinatorialValues(1, 2, 6, 13)] int threadCount,
+            [CombinatorialValues(101, 2000, 9999)] int messageCount,
+            [CombinatorialValues(1, 5, 50)] int batchElementCountThreshold)
+        {
+            var topicName = new TopicName("FakeProject", "FakeTopic");
+            var scheduler = new TestScheduler(threadCount);
+            TaskHelper taskHelper = scheduler.TaskHelper;
+            var clients = Enumerable.Range(0, clientCount).Select(_ => new FakePublisher(scheduler, taskHelper)).ToArray();
+            var settings = MakeSettings(clients, scheduler, batchElementCountThreshold: batchElementCountThreshold, batchRequestByteThreshold: 10000);
+            int shutdownCount = 0;
+            var pub = new HiPerfPublisherImpl(topicName, settings, () =>
+            {
+                shutdownCount += 1;
+                return Task.FromResult(0);
+            }, taskHelper);
+            scheduler.Run(async () =>
+            {
+                Task<string>[] tasks = new Task<string>[messageCount];
```

Comment:
Done.
I felt slightly uneasy using the more functional constructs (ie `Select`) with the imperative nature of `PublishAsync`.